File: arvores_binarias/arvore.py
class No:
    def __init__(self, valor):
        self.valor = valor
        self.esquerda = None
        self.direita = None

class ArvoreBinaria:
    def __init__(self):
        self.raiz = None

    def inserir(self, valor):
        self.raiz = self._inserir_recursivo(self.raiz, valor)

    def _inserir_recursivo(self, no, valor):
        if no is None:
            return No(valor)
        if valor < no.valor:
            no.esquerda = self._inserir_recursivo(no.esquerda, valor)
        else:
            no.direita = self._inserir_recursivo(no.direita, valor)
        return no

    def in_order(self):
        self._in_order_recursivo(self.raiz)
        print()

    def _in_order_recursivo(self, no):
        if no is not None:
            self._in_order_recursivo(no.esquerda)
            print(no.valor, end=" ")
            self._in_order_recursivo(no.direita)

    def pre_order(self):
        self._pre_order_recursivo(self.raiz)
        print()

    def _pre_order_recursivo(self, no):
        if no is not None:
            print(no.valor, end=" ")
            self._pre_order_recursivo(no.esquerda)
            self._pre_order_recursivo(no.direita)

    def post_order(self):
        self._post_order_recursivo(self.raiz)
        print()

    def _post_order_recursivo(self, no):
        if no is not None:
            self._post_order_recursivo(no.esquerda)
            self._post_order_recursivo(no.direita)
            print(no.valor, end=" ")

# Testando
arvore = ArvoreBinaria()
valores = [8, 3, 10, 1, 6, 14, 4]

for v in valores:
    arvore.inserir(v)

print("In-order:")
arvore.in_order()

print("Pré-ordem:")
arvore.pre_order()

print("Pós-ordem:")
arvore.post_order()

File: arvores_binarias/arvore.c
// Uma árvore binária é uma estrutura onde:
// Cada nó tem no máximo dois filhos: um à esquerda (left) e outro à direita (right).
// O primeiro nó é chamado de raiz.
// Nós sem filhos são chamados de folhas.
// Os valores à esquerda são menores que o valor do nó.
// Os valores à direita são maiores que o valor do nó.

// Aplicações:
// Representar hierarquias (como pastas).
// Estruturas eficientes de busca (ex: árvore binária de busca).
// Compiladores, inteligência artificial, banco de dados...

#include <stdio.h>
#include <stdlib.h>

// Estrutura do nó da árvore
typedef struct No {
    int valor;
    struct No* esquerda;
    struct No* direita;
} No;

// Cria um novo nó
No* criar_no(int valor) {
    No* novo = malloc(sizeof(No));
    novo->valor = valor;
    novo->esquerda = NULL;
    novo->direita = NULL;
    return novo;
}

// Insere um valor na árvore (binária de busca)
No* inserir(No* raiz, int valor) {
    if (raiz == NULL) {
        return criar_no(valor);
    }

    if (valor < raiz->valor) {
        raiz->esquerda = inserir(raiz->esquerda, valor);
    } else {
        raiz->direita = inserir(raiz->direita, valor);
    }

    return raiz;
}

// Percurso in-order (esquerda, raiz, direita)
void in_order(No* raiz) {
    if (raiz != NULL) {
        in_order(raiz->esquerda);
        printf("%d ", raiz->valor);
        in_order(raiz->direita);
    }
}

// Percurso pré-ordem (raiz, esquerda, direita)
void pre_order(No* raiz) {
    if (raiz != NULL) {
        printf("%d ", raiz->valor);
        pre_order(raiz->esquerda);
        pre_order(raiz->direita);
    }
}

// Percurso pós-ordem (esquerda, direita, raiz)
void post_order(No* raiz) {
    if (raiz != NULL) {
        post_order(raiz->esquerda);
        post_order(raiz->direita);
        printf("%d ", raiz->valor);
    }
}

int main() {
    No* raiz = NULL;

    raiz = inserir(raiz, 8);
    raiz = inserir(raiz, 3);
    raiz = inserir(raiz, 10);
    raiz = inserir(raiz, 1);
    raiz = inserir(raiz, 6);
    raiz = inserir(raiz, 14);
    raiz = inserir(raiz, 4);

    printf("In-order: ");
    in_order(raiz);
    printf("\n");

    printf("Pré-ordem: ");
    pre_order(raiz);
    printf("\n");

    printf("Pós-ordem: ");
    post_order(raiz);
    printf("\n");

    return 0;
}
